### PR TITLE
Stream-based Access to Matching Units

### PR DESCRIPTION
**Pull request by occamsystems**
I tried to upgrade to 6.1 and hit a performance bottleneck with Qudt#derivedUnitsFromFactorUnits. It looks like you are already working on a replacement based on the 6.2 deprecation notice. I added a stream-based accessor that is optimized for performance, especially when you only care about the best result, rather than all possible results.  

I also added a small upgrade to the pom that lets spotless run on my system with less headaches.

### CODE REVIEW

**Comment by fkleedorfer** [comment]
The performance problem was fixed by replacing the match scoring logic completely. The current approach can be optimized but it is already a lot better than before (NP hard vs logarithmic)

**Comment by fkleedorfer** [comment]
Can you confirm you are still encountering performance problems with the current main branch? Please let me know which factors cause it if it is still the case.

**Comment by occamsystems** [comment]
The new version list-based you have there is already performant enough that I could use it based on the test case that was not returning with 6.1. But I only need the first match, so it’s more convenient for me to have the stream version where I can find first without doing the intermediate list allocation.

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by occamsystems** [comment]
Also, I didn’t look at the history for the file, but the deprecated version also used to be computable. (Not 100% sure when this changed.) It might be worth looking at a reversion of that method to the previous implementation if it’s that simple. Probably not worth it if the changed happened in some other method several layers of call stack away.

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by fkleedorfer** [comment]
Just for some background: 

The complexity of the match scoring algorithm started to become noticeable with 6.1.0 because I added a lot of relationships between units that were previously unrelated (in the now misnamed [si-base-units.ttl](qudtlib-data-gen/src/main/resources/si-base-units.ttl), and I changed the behaviour or Unit.hasFactorUnits() to return true if they had only one factor unit with scaleFactor 1, referencing a different unit. That led to an explosion of the result of `FactoUnit.getAllPossibleFactorUnitCombinations()`, which was used by the match scoring algorithm, which was a naive solution to the assignment problem.

The new scoring algorithm does not rely on  `FactoUnit.getAllPossibleFactorUnitCombinations()` (which we should deprecate anyway), and I think its ineherent complexity is linear with the elements in the factor unit trees the participating units.

**Comment by fkleedorfer** [comment]
Finally I've found the time to do a more thorough review. I welcome the idea to make this functionality more performant - and I am very thankful for the work you've put into it! 

I have a few remarks on the concrete implementation:

* We now have two implementations of the ranking algorithm and there is no guarantee they do the same. There are many tests for the old method and only one test for the new. We should instead make sure the streaming method replaces the current algorithm and is thus being tested by all existing tests.
* The new streaming method returns a sorted stream, and as far as I know, there is no guarantee it will have optimal performance if only the first element is obtained. I think it would be better to defer the sorting and use Stream.sorted() where sorting is needed and Stream.min() where we need only the first.
* I am not convinced that the method of computing a `difference` value for each unit and then comparing those values is actually more efficient. The 'old' version returns a comparison value as soon as a difference is found, the suggested algorithm has to go through all the calculations in every case, for every matching unit. Moreover, I am not sure that it matters much either way because in each case, there will only be a handful of matching units (the worst case is qkdv:A0E0L0I0M0H0T0D1: 291, then qkdv:A0E0L3I0M0H0T-1D0: 96, then qkdv:A0E0L2I0M1H0T-2D0: 68 (see the [complete list](https://gist.github.com/fkleedorfer/8a00c9e9c205b00d06ff929a4d5bd962) and the [code for it](https://gist.github.com/fkleedorfer/871e785944f76ecdb43891edd121e5a9) in case you want to play with it) . Have you done any performance measuring? If so, please include it in the test classes. An ignored test will do nicely, unless you have a better idea.
* Expanding on the question of performance, it may be woth looking into caching the result of `Unit.normalize()` (or `FactorUnit.normalize()`), as this is calculated for **every** QUDT unit **every time** the operation in question is called - of course only if it really does matter for performance, as it would bloat the model.
* Another performance related idea would be to change the `Unit.matches()` algorithm such that it first compares the dimension vectors for equality. If the Unit or the FactorUnits object does not have a dimension vector, compare the normalized factors. If the dimension vectors differ, return false, otherwise compare the normalized factors. Together with caching the FactorUnit's dimension vector, this might make a difference as well.
* If we do need it, I would prefer `UnitDissimilarity` for the class name instead of `UnitDifference`. If I understand correctly, a lower value means a higher similarity. Also, the `UnitDissimilarity` class should be in a sub-package of `io.github.qudtlib.support`, such as `unitcomparison` or somesuch

**Comment by occamsystems** [comment]
So, I think if you are willing to explore more dramatic departure from the current implementation, my real inclination would be to sort units once, then use the stable sorting order as the traversal order, regardless of the FactorUnits compared. I.E., use a TreeMap that does a sorting based on unit "simplicity." This moves all of the computation burden to a single up-front sorting operation, making all other operations much faster.

WRT the match function, I did not look at its implementation, but I have found it a minor inconvenience that qudtlib generally  works with the dimension vectors on as an iri string, rather than a normal Object. (In fact, it wasn't until I started working on PRs here that I realized these were semantic iri's, which could be operated on meaningfully.) If you are open to the idea, I would definitely be happy to expand the usage of DimensionVector outside of FactorUnits to also be used on Unit and QuantityKind, which would open up a lot of avenues for simple convertability checking.

I also noticed, but did not have time to investigate, that I seemed to be getting some false positives for unit equivalence in 6.1. For example, I think Inch and Foot are coming back as equivalent units. This may, however, be due to some issue with the way my code is consuming the API, rather than a bug in qudt-lib itself.

**Comment by fkleedorfer** [comment]
> So, I think if you are willing to explore more dramatic departure from the current implementation, my real inclination would be to sort units once, then use the stable sorting order as the traversal order, regardless of the FactorUnits compared. I.E., use a TreeMap that does a sorting based on unit "simplicity." This moves all of the computation burden to a single up-front sorting operation, making all other operations much faster.

This is only possible with a specific interpretation of a 'matching' unit. My interpretation has been, so far, that the unit is most similar to the query if its factors are most similar. If I query N m / m, I want to get N-M-PER-M (maybe J-PER-M), but not N as the best match. 

I realize now that the current algorithm does not do that any more, which is a bit frustrating. For the above case, it returns N. Seems I forgot about that in a late night coding session. The relevant bit is where we check whether or not the unit name contains a hyphen, which I like, but it's probably not the good solution I thought it was. Needs more work it seems.

However, if we make the change you suggest in the quoted message above, we lose the ability to sort the result according to best fit to the query - that was not the intention. If indeed you want a different interpretation (the one that is compatible with a total static ordering of all units), I'm not against adding that as a feature. I would expose it via one or two additional matching modes. (e.g. MOST_SIMPLE_MATCH, ORDERED_BY_SIMPLICITY). 
What do you think?

**Comment by occamsystems** [comment]
Ok. I definitely hadn't captured the intended semantics there, since I was just looking at the code. I do wonder if simplestMatchingUnit/closestMatchingUnit being a pair of functions, as you suggest, is perhaps the best solution. I can definitely see applications for both.

**Comment by fkleedorfer** [comment]
> I also noticed, but did not have time to investigate, that I seemed to be getting some false positives for unit equivalence in 6.1. For example, I think Inch and Foot are coming back as equivalent units. This may, however, be due to some issue with the way my code is consuming the API, rather than a bug in qudt-lib itself.

If you find a concrete case, let's investigate. For the example, I get:
```
System.out.println(String.format("FT matches IN: %b", FT.matches(IN,1 )));
--> FT matches IN: false
```

**Comment by fkleedorfer** [comment]
> WRT the match function, I did not look at its implementation, but I have found it a minor inconvenience that qudtlib generally works with the dimension vectors as an iri string, rather than a normal Object. (In fact, it wasn't until I started working on PRs here that I realized these were semantic iri's, which could be operated on meaningfully.) If you are open to the idea, I would definitely be happy to expand the usage of DimensionVector outside of FactorUnits to also be used on Unit and QuantityKind, which would open up a lot of avenues for simple convertability checking.

Can you tell me more about the intended functionality enabled by that change? 

More Background: I am trying to keep qudtlib-java and qudtlib-js compatible in terms of model structure and compatibility. Qudtlib-java is a bit ahead at this point, and this change would push it even further. Also, I obtain the dimension vectors from QUDT, and I think some are missing. Where QUDT uses a unit only in a derived unit or a scaled unit, but does not contain the base unit for it, I generate the 'missing' unit but it is underspecified, for example, the dimension vector is missing. Example: `unit:KAT` is in QUDT, but its scalings, e.g. `unit:MilliKAT` are not. However, `unit:MilliKAT-PER-L` is in QUDT. I am an active contributor to QUDT, and could make these contributions as well, but I have to pace myself with respect to other responsibilities. The result is, anyway: not all units have dimension vectors (hence it's optional). Not in all cases will it be possible to infer one, though in some, we've reached the point now (scalings of KAT being one of them). This situation will not change if we use the DimensionVector class instead of (or in addition to) the String variant. It might be something to consider with respect to the functionality you want - there will be 'holes' in the domain of units over which that functionality is defined.

**Comment by occamsystems** [comment]
I have created a commit for the use of DimensionVector. Basically, this has three benefits:

1. It encapsulates logic about dimensionality semantics, which makes it more clear that iri semantics exist.
2. It allows us to do math to find "complementary" units. For example, it I have a velocity, and I need an acceleration, what Quantity Kinds might I be looking for to complete that?
3. It also gives a good, single point to refactor such that fractional dimensions are supported. I can actually see in the logs for my branch that such dimensions exist and are not currently handled:

WARNING: Cannot process dimension vector iri http://qudt.org/vocab/dimensionvector/A0E0L-0pt5I0M1H0T-2D0: unexpected number of dimensions: 11
Jan 15, 2024 10:28:01 AM io.github.qudtlib.model.DimensionVector <init>
WARNING: Cannot process dimension vector iri http://qudt.org/vocab/dimensionvector/A0E0L-0pt5I0M1H0T-2D0: unexpected number of dimensions: 11
Jan 15, 2024 10:28:01 AM io.github.qudtlib.model.DimensionVector <init>
WARNING: Cannot process dimension vector iri http://qudt.org/vocab/dimensionvector/A0E0L-0pt5I0M1H0T-2D0: unexpected number of dimensions: 11

**Comment by fkleedorfer** [comment]
Please put the dimensionvector stuff on a new PR. It's orthogonal to the unit matching PR. I don't hate it though.

**Comment by occamsystems** [comment]
I can do that. I also went ahead and did the refactor to support non-integer dimension vectors.

File: qudtlib-model/src/main/java/io/github/qudtlib/model/DimensionVector.java
Comment (fkleedorfer) [inline qudtlib-model/src/main/java/io/github/qudtlib/model/DimensionVector.java:1]: Looks like you are not generating the fractional dimension according to the QUDT conversion (0pt5) when generating the dim vector iri

**Comment by occamsystems** [comment]
Yes, it just occurred to me that I need to use 0.# instead of #.# as the decimal format. I took the afternoon off, since it’s a holiday in the US today. I will make a new branch tomorrow, cherry pick those commits over, and fix the issue.

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by fkleedorfer** [comment]
Needs more work - reopen if interested

**Pull request closed without merging**